src/vdaf.rs

```diff
@@ -240,6 +242,18 @@ impl<F> AsRef<[F]> for OutputShare<F> {
     }
 }
 
+impl<F: Encode> Encode for OutputShare<F> {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        encode_u16_items(bytes, &(), &self.0)
```

Comment:
It makes sense to avoid the length prefix if we can, but I want to make sure I understand the suggestion.

AFAICT, the expected length of an output share/aggregate share will vary from VDAF to VDAF, and is not exposed by any existing VDAF methods. (e.g. prio3 output shares & aggregate shares will both have a length based on the `flp::Type` in use; for poplar1, output shares & aggregate shares will have a length based on the number of candidate prefixes in the aggregation param. FWIW prio3 does have an `output_len` method, but this is on the `prio3`-specific struct and not the Vdaf trait)

Is there a way for a user of this library to determine the expected length of an output share/aggregation share to pass to `decode_with_param`? Or is the suggestion that, if implementations want to deserialize these structures, they should serialize the additional information required for decoding themselves in a "wrapper" message around what is encoded by these `Encode` implementations?

(BTW, I found only one place that was calling `AggregateShare::decode_with_param`, and it is prio3-specific in that it assumes that the aggregate share length is equal to the output length & that those lengths are constant based on the VDAF [https://github.com/divviup/ppm-prototype/blob/40fcb1abe6b9480bfaa5beec56bf0d6fce474473/src/bin/collector.rs#L16]. So we may not be OK with `AggregateShare` as it stands!)

Reply:
Gentle ping @cjpatton. I'd like to resolve this.

After discussion with our group, two solutions come to mind:
1) Drop the `usize` decoding parameter entirely, logically decoding with no parameter. Have `AggregateShare` & `OutputShare` implement `TryFrom<&[u8]>` and `TryInto<Vec<u8>>`. (`Encode`/`Decode` are not appropriate because they operate on streams of values, while in this scenario we'd be decoding based on the total length of the vector, since we aren't encoding any kind of length information.)

2) Extend the `Vdaf` trait enough to have associated types for "Aggregate Share Decoding Param" & "Output Share Decoding Param", along with methods to retrieve the appropriate parameters. This would allow VDAFs more control over the decoding process, at the cost of complxifying the `Vdaf` trait even more.

Thoughts?

Reply:
I think I see the problem: You can't decode an out or agg share without knowing the expected length, and right now there's nothing generic about `Vdaf` that tells you what length to expect. (1) works because `AggregateShare` and `OutputShare` are generic in the `FieldElement`, which specifies a a fixed length for each field element. (2) works because it requires the VDAF to specify this length.

My preference would be (1) because I think it's best to avoid restraining the VDAF API as much as possible, in particular for behaviors that are unspecified.

Reply:
Yep, that's exactly the problem.

Following your suggestion, I switched from Encode/Decode to From/Into<Vec<u8>>. The trait bounds required to make From/Into work on _references to_ the shares rather than the shares themselves are a little funky, but this is cleaner.

I left the bounds off of Vdaf::OutputShare, but updated the bounds on Vdaf::AggregateShare. This makes sense to me since an aggregate share _must_ be transmitted between aggregators (and so must be serialized), while an output share is not transmitted.